volumina/utility/preferences.py

```diff
@@ -0,0 +1,143 @@
+"""
+
+import logging
+import os
+import pathlib
+import pickle
+import threading
+from typing import Any, Tuple, MutableMapping
+
+logger = logging.getLogger(__name__)
+
+
+class _Preferences:
+    def __init__(self, path: pathlib.Path):
+        self._path = path
+        self._data = _load_preferences(path)
+        self._lock = threading.Lock()
+
+    def get(self, group: str, setting: str, default: Any = None) -> Any:
+        return self.getmany((group, setting, default))[0]
+
+    def set(self, group: str, setting: str, value: Any) -> None:
+        self.setmany((group, setting, value))
+
+    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+        with self._lock:
+            result = []
+            for group, setting, default in args:
+                try:
+                    result.append(self._data[group][setting])
+                except KeyError:
+                    result.append(default)
+            return tuple(result)
+
+    def setmany(self, *args: Tuple[str, str, Any]) -> None:
+        with self._lock:
+            for group, setting, value in args:
+                self._data.setdefault(group, {})[setting] = value
+            try:
+                with open(self._path, "wb") as f:
+                    # Use the lowest available protocol for backwards compatibility.
+                    pickle.dump(self._data, f, protocol=0)
+            except Exception as e:
+                logger.exception(f"Failed to save preferences to {str(self._path)!r}")
+                if not isinstance(e, IOError):
+                    self._path.unlink()
+
+    def set_location(self, path: pathlib.Path) -> None:
+        with self._lock:
+            self._path = path
+            self._data = _load_preferences(path)
+
+
+def _load_preferences(path: pathlib.Path) -> MutableMapping[str, MutableMapping[str, Any]]:
+    try:
+        with open(path, "rb") as f:
+            return pickle.load(f)
+    except FileNotFoundError:
+        return {}
+    except Exception as e:
+        logger.exception(f"Failed to load preferences from {str(path)!r}")
+        if not isinstance(e, IOError):
+            path.unlink()
+        return {}
+
+
+_preferences = _Preferences(pathlib.Path.home() / ".ilastik_preferences")
+
+
+def get(group: str, setting: str, default: Any = None) -> Any:
+    """Return a preference setting in the specified group.
+
+    If there is no such group, or there is no such setting in the
+    existing group, the default value is returned.
+
+    See Also:
+        :func:`getmany`.
+    """
+    return _preferences.get(group, setting, default)
+
+
+def set(group: str, setting: str, value: Any) -> None:
+    """Set a preference setting in the specified group.
+
+    If the group or the setting in a group does not exist, it will be
+    created.
+
+    See Also:
+        :func:`setmany`.
+    """
+    return _preferences.set(group, setting, value)
+
+
+def getmany(*args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
```

Comment:
I think as no other module except tests uses these methods (`getmany`, `setmany`), we should not make them public nor on module level, nor on class.

Reply:
`setmany` is [used in liastik](https://github.com/ilastik/ilastik/pull/2135/files#diff-870f52209923664e0309e4912a5f5069R507).

Reply:
Do we need a full function to avoid calling two consecutive sets?

Reply:
Yes, because another thread might change one of the values between two calls.